### DISCUSSION TITLE
Settings in Pcp_combine/derive conf file

### DISCUSSION THREAD

**Discussion by BinbinZhou-NOAA** [dtcenter/METplus#1831] [Configuration] [answered]
I have an issue on running Pcp_combine -derive in METplus. 
I want to average 24 fhr sea ice concentration (name=ICEC, level=Z0) from GEFS.
My first test is using  GEFS member#1 forecast  initiated at 00Z on 20220918. 
The ICEC values from following 4  output files should be averaged:
  
gefs.ens01.t00z.grid3.f006.grib2
gefs.ens01.t00z.grid3.f012.grib2
gefs.ens01.t00z.grid3.f018.grib2
gefs.ens01.t00z.grid3.f024.grib2

Following the example of use case of Pcp_combine -derive, my conf file likes below:
#########################################
[config]
PROCESS_LIST = PCPCombine

LOOP_BY = INIT
INIT_TIME_FMT = %Y%m%d%H
INIT_BEG =  2022091800
INIT_END = 2022091800
INIT_INCREMENT = 1H

LEAD_SEQ = 24H
LOOP_ORDER = times

FCST_PCP_COMBINE_RUN = True
FCST_PCP_COMBINE_METHOD = DERIVE

FCST_PCP_COMBINE_INPUT_DIR = {ENV[modelpath]}
FCST_PCP_COMBINE_INPUT_TEMPLATE = gefs.ens01.t{init?fmt=%H}z.grid3.f{lead?fmt=%3H}.grib2

FCST_PCP_COMBINE_OUTPUT_DIR = {OUTPUT_BASE}
FCST_PCP_COMBINE_OUTPUT_TEMPLATE = gefs.ens01.t{init?fmt=%H}z.grid3.icec_24h.f{lead?fmt=%3H}.nc

FCST_PCP_COMBINE_STAT_LIST = mean
FCST_PCP_COMBINE_DERIVE_LOOKBACK = 24H

FCST_PCP_COMBINE_MIN_FORECAST = 6H
FCST_PCP_COMBINE_MAX_FORECAST = 2d

FCST_IS_PROB = false

FCST_PCP_COMBINE_INPUT_DATATYPE = GRIB

FCST_PCP_COMBINE_INPUT_ACCUMS = 6H
FCST_PCP_COMBINE_INPUT_NAMES = ICEC
FCST_PCP_COMBINE_INPUT_LEVELS = Z0
FCST_PCP_COMBINE_INPUT_OPTIONS =

FCST_PCP_COMBINE_OUTPUT_ACCUM = 24H
FCST_PCP_COMBINE_OUTPUT_NAME =
######################################################

The running is successful with output a netcdf file. But checking the log file, showing below message
DEBUG 2: Performing derivation command (mean) for 4 files.
DEBUG 1: Reading data (name="ICEC"; level="Z0";) from input file: atmos.20220918/gefs/gefs.ens01.t18z.grid3.f006.grib2
DEBUG 1: Reading data (name="ICEC"; level="Z0";) from input file: atmos.20220918/gefs/gefs.ens01.t12z.grid3.f006.grib2
DEBUG 1: Reading data (name="ICEC"; level="Z0";) from input file: atmos.20220918/gefs/gefs.ens01.t06z.grid3.f006.grib2
DEBUG 1: Reading data (name="ICEC"; level="Z0";) from input file: atmos.20220918/gefs/gefs.ens01.t00z.grid3.f006.grib2

This means the searching look up wrong files: accumulation does not on lead times 006, 012, 018 and 024, but on initial
times (00,06,12,18) instead. 

I think the issue is in the time settings in my conf file, but don't know where is.

Binbin

**Comment by j-opatz** [comment]
Hi Binbin,

And thank you for your question. After reviewing your configuration file (thank you for providing that!) and the output from the example PCPCombine output that uses the `-derive` logic, I'm going to request that @georgemccabe take a look at your issue when he's in the office, as he is more knowledgeable on why this behavior is happening and how to get the output you're looking for. 

If this follows the example case logic, the first file search of your PCPCombine use case should be looking for a valid time of 2022-09-19 at 00:00:00 between 9 hours and 2 days (your `FCST_PCP_COMBINE_MIN_FORECAST` and `FCST_PCP_COMBINE_MAX_FORECAST` bounds) with 1 hour intervals. During this first search, the lowest lead time would correspond to a file named

`gefs.ens01.t15z.grid3.f009.grib2`

following the template of your other files (initialized at 2022-09-18 at 15:00:00). If it can't find that file, it should slowly expand the lead time while simultaneously keeping the valid time at 2022-09-19 at 00:00:00. It follows that the initialization time must retrograde. So the next files it searches for are 
```
gefs.ens01.t14z.grid3.f010.grib2
gefs.ens01.t13z.grid3.f011.grib2
gefs.ens01.t12z.grid3.f012.grib2
```

and so on. So the initialization times *should* be iterated over, as should the lead times. 

But the idea that the first file it's finding in it's search is 

`gefs.ens01.t18z.grid3.f006.grib2`

would seem to me that it somehow reset the lead time (below your 9 hour limit) to f006. However, as John HG pointed out, the files it gets are correct for accumulations. So it seems like this may be an internal tool logic.

**Reply by BinbinZhou-NOAA** [reply]
The searching seems  over both initial and validation times.  To  search  various forecast hours only at fixed initial time 2022091800,
I removed all other initial times files (06, 12 and 18Z, only keep 00Z files )  in the  directory, then files read are what I want:
gefs.ens01.t00z.grid3.f006.grib2
gefs.ens01.t00z.grid3.f012.grib2
gefs.ens01.t00z.grid3.f018.grib2
gefs.ens01.t00z.grid3.f024.grib2 

The coming issue is in deriving the second day average. That is to average 
the ICEC from the forecast 30, 36,42 and 48 fhr  
gefs.ens01.t00z.grid3.f030.grib2
gefs.ens01.t00z.grid3.f036.grib2
gefs.ens01.t00z.grid3.f042.grib2
gefs.ens01.t00z.grid3.f048.grib2

In the conf file, set
LEAD_SEQ = 48H
and all other settings are kept same

The running is fine, but log file shows still reading 
the first forecast day's 006, 012 018 and 024 files, not the second day's 4 files.

Not sure any other settings need to re-set

**Comment by georgemccabe** [comment, marked as answer]
Hi @BinbinZhou-NOAA, I believe this is what you want to add to accomplish what you need:

`FCST_PCP_COMBINE_CONSTANT_INIT = True`

This keeps the initialization time constant while finding the files to process. Please let me know if this does not work for you.

**Reply by BinbinZhou-NOAA** [reply]
George,

   Yes, after adding this new setting, it reads correct files. Thanks!
It is better to add this important setting in the Pcp_combine  use-case examples in the next version of  METplus user guide.

**Reply by georgemccabe** [reply]
Glad to hear this worked for you. I created issue #1835 to improve the documentation to include clear information about this setting.

**Reply by BinbinZhou-NOAA** [reply]
Thanks for this update!

Binbin
